base/multidimensional.jl

```diff
@@ -216,70 +219,100 @@ end
-@inline index_shape_dim(inds,             ::Colon)          = (OneTo(trailingsize(inds)),)
-@inline index_shape_dim(inds,             ::Colon, i::CI0, I::CI0...) =
-    (OneTo(trailingsize(inds)), index_shape_dim((), i, I...)...)
-@inline function index_shape_dim(inds,    ::Colon, i, I...)
-    inds1, indstail = IteratorsMD.split(inds, Val{1})
-    (_gimme_a_range(inds1...), index_shape_dim(indstail, i, I...)...)
-end
-@inline index_shape_dim(inds,    ::Real...)             = ()
-@inline index_shape_dim(inds,    ::Real, I...)          = index_shape_dim(safe_tail(inds), I...)
-@inline index_shape_dim{N}(inds, ::CartesianIndex{N}, I...) =
-    index_shape_dim(IteratorsMD.split(inds, Val{N})[2], I...)
-@inline index_shape_dim(inds, i::AbstractArray, I...) =
-    (indices(i)..., index_shape_dim(safe_tail(inds), I...)...)
-@inline index_shape_dim(inds, i::AbstractArray{Bool}, I...) =
-    (OneTo(sum(i)), index_shape_dim(safe_tail(inds), I...)...)
-# single CartesianIndex version not needed because of call to flatten in _getindex...
-# ...but array of CartesianIndex is not covered
-@inline function index_shape_dim{N}(inds, i::AbstractArray{CartesianIndex{N}}, I...)
-    indsN, indstail = IteratorsMD.split(inds, Val{N})
-    (indices(i)..., index_shape_dim(indstail, I...)...)
-end
-_gimme_a_range() = OneTo(1)
-_gimme_a_range(inds::Range) = inds
-
-# Convert Colon indices into explicit indices
-@inline decolon(A::AbstractArray, ::Colon) = (linearindices(A),)
-@inline decolon(A::AbstractArray, I...) = decolon_dim(indices(A), I...)
-@inline decolon_dim(inds)  = ()
-@inline decolon_dim(inds::Tuple{Any}, ::Colon)       = inds
-@inline decolon_dim(inds,             ::Colon)       = (OneTo(trailingsize(inds)),)
-@inline decolon_dim(inds,             ::Colon, i::CI0, I::CI0...) = (OneTo(trailingsize(inds)), i, I...)
-@inline function decolon_dim(inds,    ::Colon, I...)
-    inds1, indstail = IteratorsMD.split(inds, Val{1})
-    (maybe_oneto(inds1...), decolon_dim(indstail, I...)...)
-end
-@inline decolon_dim(inds, i1, I...) = (i1, decolon_dim(safe_tail(inds), I...)...)
-@inline function decolon_dim{N}(inds, i1::AbstractArray{CartesianIndex{N}}, I...)
-    indsN, indstail = IteratorsMD.split(inds, Val{N})
-    (i1, decolon_dim(indstail, I...)...)
-end
-maybe_oneto(i) = i
-maybe_oneto() = OneTo(1)
+index_shape() = ()
+@inline index_shape(::Real, rest...) = index_shape(rest...)
+@inline index_shape(A::AbstractArray, rest...) = (indices(A)..., index_shape(rest...)...)
+
+"""
+    LogicalIndex(mask)
+
+The `LogicalIndex` type is a special vector that simply contains all indices I
+where `mask[I]` is true. This specialized type does not support indexing
+directly as doing so would require O(n) lookup time. `AbstractArray{Bool}` are
+wrapped with `LogicalIndex` upon calling `to_indices`.
+"""
+immutable LogicalIndex{T, A<:AbstractArray{Bool}} <: AbstractVector{T}
+    mask::A
+    sum::Int
+    LogicalIndex(mask::A) = new(mask, countnz(mask))
+end
```

Comment:
Yes; in general I was surprised how hard I had to fight to match the allocating `find` algorithm.  `find(::BitArray)` is wonderfully optimized and tough to beat.